overlord/snapstate/handlers.go

```diff
@@ -1166,29 +1241,47 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) (err error) {
 		return copyDataErr
 	}
 
-	action := triggeredMigration(newInfo.Base, opts)
-	var undo func()
+	var oldBase string
+	if oldInfo != nil {
+		oldBase = oldInfo.Base
+	}
 
-	switch action {
+	snapName := snapsup.InstanceName()
+	switch triggeredMigration(oldBase, newInfo.Base, opts) {
 	case hidden:
-		undo, err = m.doHiddenMigration(t, st, snapsup, newInfo)
+		if err := m.backend.HideSnapData(snapName); err != nil {
+			return err
+		}
+
+		snapsup.MigratedHidden = true
 	case revertHidden:
-		undo, err = m.revertHiddenMigration(t, st, snapsup)
+		if err := m.backend.UndoHideSnapData(snapName); err != nil {
+			return err
+		}
+
+		snapsup.UndidHiddenMigration = true
 	case home:
-		undo, err = m.doHomeMigration(t, st, snapsup, newInfo.Revision)
+		if err := m.backend.InitExposedSnapHome(snapName, newInfo.Revision); err != nil {
+			return err
+		}
+
+		snapsup.MigratedToExposedHome = true
```

Comment:
given how this one is used in undo, shouldn't it set depending whether the directory existed already or not? but then we should use EnableExposedHome = true as well ?

Reply:
The current state is functionally equivalent to the previous version (just less complex) so this isn't making the distinction between enable/create yet. But in the end, yes, this will have extra logic to do an enable or a create and set the flags accordingly